### ISSUE TITLE
Trying to backward through the graph a second time, but the saved intermediate results have already been freed. Specify retain_graph=True when calling backward the first time.

### ISSUE THREAD

**Issue by WisleyWang** [sgadgil6/cnslab_fmri#4]
In this run_training_lstm.py,
when running to  loss.backward()  ,break

if I use retain_graph=True ,then buff will OOM

**Comment by sgadgil6** [comment]
Can you do a pull request with the fix?

**Comment by WisleyWang** [comment]
> Can you do a PR with the fix?

class fMRI_LSTM(nn.Module):
    def __init__(self, hidden_dim, input_dim):
        super(fMRI_LSTM, self).__init__()
        self.hidden_dim = hidden_dim
        self.lstm = nn.LSTM(input_dim, hidden_dim,num_layers=1, batch_first=True) # 
        self.linear = nn.Linear(hidden_dim,1)
        self.hidden= self.init_hidden()
        self.dropout = nn.Dropout(p=0.2)
        # self.target_size = target_size
        self.activate=nn.Sigmoid()

    def init_hidden(self, ):#
        return (torch.randn(1, 64, self.hidden_dim).to(device), torch.randn(1,64, self.hidden_dim).to(device))

    def forward(self, x):
        #x = x.T
        lstm_out, self.hidden = self.lstm(x)
        lstm_out = lstm_out.squeeze()[:, -1, :] # here ,you don't need to pass hidden layer state
        out = self.dropout(lstm_out)
        # mean_pooling_output = F.avg_pool1d(out,)
        linear_output = self.linear(out)
        #linear_output = self.dropo(linear_output)
        
        final_output = self.activate(linear_output)
        return final_output

**Comment by WisleyWang** [comment]
> Can you do a PR with the fix?

###
class fMRI_LSTM(nn.Module):
    def __init__(self, hidden_dim, input_dim):
        super(fMRI_LSTM, self).__init__()
        self.hidden_dim = hidden_dim
        self.lstm = nn.LSTM(input_dim, hidden_dim,num_layers=1, batch_first=True) # 这里默认隐藏层为1
        self.linear = nn.Linear(hidden_dim,1)
        self.hidden= self.init_hidden()
        self.dropout = nn.Dropout(p=0.2)
        # self.target_size = target_size
        self.activate=nn.Sigmoid()

    def init_hidden(self, ):# 隐藏层的初始参数，第一个维度代表的层数
        return (torch.randn(1, 64, self.hidden_dim).to(device), torch.randn(1,64, self.hidden_dim).to(device))

    def forward(self, x):
        #x = x.T
        lstm_out, self.hidden = self.lstm(x)
        lstm_out = lstm_out.squeeze()[:, -1, :] # 只取了最后一个状态量
        out = self.dropout(lstm_out)
        # mean_pooling_output = F.avg_pool1d(out,)
        linear_output = self.linear(out)
        #linear_output = self.dropo(linear_output)
        
        final_output = self.activate(linear_output)
        return final_output

**Comment by WisleyWang** [comment]
OK

**Comment by sgadgil6** [comment]
Thank you
